### PR TITLE
Activate discount, generate qr code, additional field in discount - promocode

### PR DESCRIPTION
**Pull request by Bozhokmaria**

### CODE REVIEW

File: src/main/java/com/exadel/discountwebapp/user_discount/controller/UserDiscountController.java [hunk 1]
```diff
@@ -0,0 +1,30 @@
+package com.exadel.discountwebapp.user_discount.controller;
```
Comment (ahlukhouski2) [inline src/main/java/com/exadel/discountwebapp/user_discount/controller/UserDiscountController.java]: let's rename package to com.exadel.discountwebapp.userdiscount.controller (without underscore)

File: src/main/java/com/exadel/discountwebapp/user_discount/controller/UserDiscountController.java [hunk 2]
```diff
@@ -0,0 +1,30 @@
...
+
+@RestController
+@RequiredArgsConstructor
+@RequestMapping("/api/user-discount")
+public class UserDiscountController {
+
+    private final UserDiscountService userDiscountService;
+
+    @PostMapping(produces = MediaType.IMAGE_PNG_VALUE)
+    public byte[] create(@Valid @RequestBody UserDiscountRequestVO request) throws IOException, WriterException {
```
Comment (ahlukhouski2) [inline src/main/java/com/exadel/discountwebapp/user_discount/controller/UserDiscountController.java]: Use @SneakyThrows

File: src/main/java/com/exadel/discountwebapp/user_discount/mapper/UserDiscountMapper.java [hunk 3]
```diff
@@ -0,0 +1,73 @@
...
+    }
+
+
+    private void provideUserDiscountDependencies(UserDiscountRequestVO request, UserDiscount userDiscount) throws IOException, WriterException {
+        var user = userRepository.findById(request.getUserId())
+                .orElseThrow(() -> new EntityNotFoundException(User.class, "id", request.getUserId()));
+        var discount = discountRepository.findById(request.getDiscountId())
+                .orElseThrow(() -> new EntityNotFoundException(Discount.class, "id", request.getDiscountId()));
+
+        String qrcodeData = "First client name: " + user.getFirstName() +
```
Comment (ahlukhouski2) [inline src/main/java/com/exadel/discountwebapp/user_discount/mapper/UserDiscountMapper.java]: Let's move the logic of QR code generation to the UserDiscount Service

File: src/main/java/com/exadel/discountwebapp/user_discount/mapper/UserDiscountMapper.java [hunk 4]
```diff
@@ -0,0 +1,73 @@
...
+                "\n Last client name: " + user.getLastName() +
```
Comment (ahlukhouski2) [inline src/main/java/com/exadel/discountwebapp/user_discount/mapper/UserDiscountMapper.java]: Let's use structured information in QR code. We can consider json:
{
     "vendorEmail":"...",
     ....
}

File: src/main/java/com/exadel/discountwebapp/user_discount/qrcode_generator/QRCodeGenerator.java [hunk 5]
```diff
@@ -0,0 +1,27 @@
...
+import org.springframework.stereotype.Component;
+
+import java.io.ByteArrayOutputStream;
+import java.io.IOException;
+
+@Data
+@Component
+public class QRCodeGenerator {
+
+    public byte[] getQRCodeImage(String text, int width, int height) throws WriterException, IOException {
```
Comment (ahlukhouski2) [inline src/main/java/com/exadel/discountwebapp/user_discount/qrcode_generator/QRCodeGenerator.java]: @SneakyThrows

File: src/main/java/com/exadel/discountwebapp/user_discount/qrcode_generator/QRCodeGenerator.java [hunk 6]
```diff
@@ -0,0 +1,27 @@
...
+import com.google.zxing.client.j2se.MatrixToImageWriter;
+import com.google.zxing.common.BitMatrix;
+import com.google.zxing.qrcode.QRCodeWriter;
+import lombok.Data;
+import org.springframework.stereotype.Component;
+
+import java.io.ByteArrayOutputStream;
+import java.io.IOException;
+
+@Data
```
Comment (ahlukhouski2) [inline src/main/java/com/exadel/discountwebapp/user_discount/qrcode_generator/QRCodeGenerator.java]: @Data?

File: src/main/java/com/exadel/discountwebapp/user_discount/mapper/UserDiscountMapper.java [hunk 7]
```diff
@@ -0,0 +1,73 @@
...
+                "\n Vendor email: " + discount.getVendor().getEmail() +
+                "\n Discount: " + discount.getTitle() +
+                "\n Discount start date: " + discount.getStartDate() +
+                "\n Discount expiration date: " + discount.getExpirationDate();
+
+        if (discount.getPromocode() != null) {
+            qrcodeData += "\n Discount promocode " + discount.getPromocode();
+        }
+
+        userDiscount.setQrcode(qrCodeGenerator.getQRCodeImage(qrcodeData, 300, 300));
```
Comment (ahlukhouski2) [inline src/main/java/com/exadel/discountwebapp/user_discount/mapper/UserDiscountMapper.java]: Move 300 to constants

File: src/main/java/com/exadel/discountwebapp/user_discount/service/UserDiscountService.java [hunk 8]
```diff
@@ -0,0 +1,43 @@
...
+@RequiredArgsConstructor
+public class UserDiscountService {
+
+    private final UserDiscountRepository userDiscountRepository;
+    private final UserDiscountMapper userDiscountMapper;
+
+    @Transactional(propagation = Propagation.REQUIRED)
+    public byte[] create(UserDiscountRequestVO request) throws IOException, WriterException {
+        UserDiscount userDiscount = userDiscountMapper.toEntity(request);
+        if (userDiscountRepository.existsById(new UserDiscount.UserDiscountId(request.getUserId(), request.getDiscountId()))) {
```
Comment (ahlukhouski2) [inline src/main/java/com/exadel/discountwebapp/user_discount/service/UserDiscountService.java]: this check should be before toEntity method

File: src/main/java/com/exadel/discountwebapp/user_discount/controller/UserDiscountController.java (hunk 1 above)
Comment (Bozhokmaria) [reply src/main/java/com/exadel/discountwebapp/user_discount/controller/UserDiscountController.java]: Done

File: src/main/java/com/exadel/discountwebapp/user_discount/controller/UserDiscountController.java (hunk 2 above)
Comment (Bozhokmaria) [reply src/main/java/com/exadel/discountwebapp/user_discount/controller/UserDiscountController.java]: Used

File: src/main/java/com/exadel/discountwebapp/user_discount/mapper/UserDiscountMapper.java (hunk 3 above)
Comment (Bozhokmaria) [reply src/main/java/com/exadel/discountwebapp/user_discount/mapper/UserDiscountMapper.java]: Done

File: src/main/java/com/exadel/discountwebapp/user_discount/mapper/UserDiscountMapper.java (hunk 4 above)
Comment (Bozhokmaria) [reply src/main/java/com/exadel/discountwebapp/user_discount/mapper/UserDiscountMapper.java]: Separate object for qrcode data+ Jason format after scanning qrcode

File: src/main/java/com/exadel/discountwebapp/user_discount/qrcode_generator/QRCodeGenerator.java (hunk 5 above)
Comment (Bozhokmaria) [reply src/main/java/com/exadel/discountwebapp/user_discount/qrcode_generator/QRCodeGenerator.java]: Done

File: src/main/java/com/exadel/discountwebapp/user_discount/qrcode_generator/QRCodeGenerator.java (hunk 6 above)
Comment (Bozhokmaria) [reply src/main/java/com/exadel/discountwebapp/user_discount/qrcode_generator/QRCodeGenerator.java]: Removed

File: src/main/java/com/exadel/discountwebapp/user_discount/mapper/UserDiscountMapper.java (hunk 7 above)
Comment (Bozhokmaria) [reply src/main/java/com/exadel/discountwebapp/user_discount/mapper/UserDiscountMapper.java]: Moved width and height constants

File: src/main/java/com/exadel/discountwebapp/user_discount/service/UserDiscountService.java (hunk 8 above)
Comment (Bozhokmaria) [reply src/main/java/com/exadel/discountwebapp/user_discount/service/UserDiscountService.java]: Done

File: src/main/java/com/exadel/discountwebapp/userdiscount/entity/UserDiscount.java [hunk 9]
```diff
@@ -35,6 +36,9 @@
     @ManyToOne
     private Discount discount;
 
+    @Column(name = "qrcode")
```
Comment (dmitriychernov7) [inline src/main/java/com/exadel/discountwebapp/userdiscount/entity/UserDiscount.java]: rename to ud_qrcode

File: src/main/java/com/exadel/discountwebapp/userdiscount/service/UserDiscountService.java [hunk 10]
```diff
@@ -0,0 +1,78 @@
...
+
+        QRCodeData qrcodeData = new QRCodeData();
+        qrcodeData.setUserFirstName(user.getFirstName());
+        qrcodeData.setUserLastName(user.getLastName());
+        qrcodeData.setVendorTitle(discount.getVendor().getTitle());
+        qrcodeData.setVendorEmail(discount.getVendor().getEmail());
+        qrcodeData.setDiscountTitle(discount.getTitle());
+        if (discount.getPromocode() != null) {
+            qrcodeData.setDiscountPromocode(discount.getPromocode());
+        }
```
Comment (dmitriychernov7) [inline src/main/java/com/exadel/discountwebapp/userdiscount/service/UserDiscountService.java]: if there is no promocode, why would we generate QR code?

File: src/main/java/com/exadel/discountwebapp/userdiscount/entity/UserDiscount.java (hunk 9 above)
Comment (Bozhokmaria) [reply src/main/java/com/exadel/discountwebapp/userdiscount/entity/UserDiscount.java]: no more qr_code column in db

File: src/main/java/com/exadel/discountwebapp/user/entity/User.java [hunk 11]
```diff
@@ -57,9 +61,9 @@
     @JoinColumn(name = "rol_id")
     private Role role;
 
-    @OneToMany(mappedBy = "user")
+    @ManyToMany(mappedBy = "users")
     @EqualsAndHashCode.Exclude
-    private List<UserDiscount> discounts = new ArrayList<>();
+    private List<Discount> discounts = new ArrayList<>();
```
Comment (dmitriychernov7) [inline src/main/java/com/exadel/discountwebapp/user/entity/User.java]: why did you change this?

File: src/main/java/com/exadel/discountwebapp/userdiscount/vo/UserDiscountResponseVO.java [hunk 12]
```diff
@@ -0,0 +1,13 @@
...
+import lombok.Data;
+import lombok.EqualsAndHashCode;
+import lombok.NoArgsConstructor;
+import lombok.experimental.SuperBuilder;
+
+@Data
+@SuperBuilder
+@NoArgsConstructor
+@EqualsAndHashCode(callSuper = true)
+public class UserDiscountResponseVO extends UserDiscountBaseVO {
```
Comment (dmitriychernov7) [inline src/main/java/com/exadel/discountwebapp/userdiscount/vo/UserDiscountResponseVO.java]: this class is not used anywhere. we don't need it

File: src/main/java/com/exadel/discountwebapp/userdiscount/vo/UserDiscountBaseVO.java [hunk 13]
```diff
@@ -0,0 +1,13 @@
...
+import lombok.Data;
+import lombok.NoArgsConstructor;
+import lombok.experimental.SuperBuilder;
+
+import java.io.Serializable;
+
+@Data
+@SuperBuilder
+@NoArgsConstructor
+public class UserDiscountBaseVO implements Serializable {
```
Comment (dmitriychernov7) [inline src/main/java/com/exadel/discountwebapp/userdiscount/vo/UserDiscountBaseVO.java]: we don't need this class

File: src/main/java/com/exadel/discountwebapp/user/entity/User.java (hunk 11 above)
Comment (Bozhokmaria) [reply src/main/java/com/exadel/discountwebapp/user/entity/User.java]: Searching with specification doesnt work without this changes 

/api/discounts?query=users.id:1
Or 
/api/users?query=discounts.id:3

File: src/main/java/com/exadel/discountwebapp/userdiscount/service/UserDiscountService.java [hunk 14]
```diff
@@ -0,0 +1,60 @@
...
+        byte[] qrcode = qrCodeGenerator.generateQRCodeImage(dataForQRCode(request), QRCODE_WIDTH, QRCODE_HEIGHT);
+        return qrcode;
+    }
+
+    @SneakyThrows
+    private String dataForQRCode(UserDiscountRequestVO request) {
+        //TODO: add valid url adress
+        //String data = "http://localhost:8080/api/qrcode/" + request.getUserId() + "/" + request.getDiscountId();
+
+        String data = "https://sandbox-discount.herokuapp.com/" + request.getUserId() + "/" + request.getDiscountId();
```
Comment (dmitriychernov7) [inline src/main/java/com/exadel/discountwebapp/userdiscount/service/UserDiscountService.java]: we need to provide url through properties

File: src/main/java/com/exadel/discountwebapp/user/entity/User.java (hunk 11 above)
Comment (dmitriychernov7) [reply src/main/java/com/exadel/discountwebapp/user/entity/User.java]: /api/discounts?query=users.user.id:1 search discounts by user id
/api/discounts?query=users.id.userId:1 search discounts by user id

File: src/main/java/com/exadel/discountwebapp/userdiscount/service/UserDiscountService.java (hunk 14 above)
Comment (dmitriychernov7) [reply src/main/java/com/exadel/discountwebapp/userdiscount/service/UserDiscountService.java]: https://sandbox-discount.herokuapp.com/qrcode?userId=.....&dicountId=....

File: src/main/java/com/exadel/discountwebapp/userdiscount/qrcodegenerator/QRCodeController.java [hunk 15]
```diff
@@ -0,0 +1,20 @@
...
+
+@RestController
+@RequestMapping("/api/qrcode")
+@RequiredArgsConstructor
+public class QRCodeController {
+
+    private final QRCodeService qrCodeService;
+
+    @GetMapping("/{userId}/{discountId}")
+    public QRCodeResponseVO getQRCodeViewPage(@PathVariable Long userId, @PathVariable Long discountId) {
```
Comment (dmitriychernov7) [inline src/main/java/com/exadel/discountwebapp/userdiscount/qrcodegenerator/QRCodeController.java:17]: @RequestParam("userId")

**Review by dmitriychernov7 — APPROVED** [review]
